### PR TITLE
feat: add val_period parameter to control validation frequency during training

### PR DESCRIPTION
**Pull request by Bovey0809**
- Add val_period parameter to default.yaml (default: 1 = every epoch)
- Register val_period as integer config key in cfg/__init__.py
- Initialize val_period in BaseTrainer.__init__
- Update training loop to conditionally run validation based on val_period
- Validation still runs on final epoch and early stopping regardless of val_period

This allows users to control how often validation runs during training, similar to save_period for checkpoint saving. Useful for reducing validation overhead during long training runs.

Example usage: val_period=5 (validate every 5 epochs)

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds a new `val_period` setting to control how often validation runs during training (e.g., every N epochs) 🧪⏱️

### 📊 Key Changes
- Introduced a new config key `val_period` across the config system ✅
  - Registered in `ultralytics/cfg/__init__.py`
  - Added to defaults in `ultralytics/cfg/default.yaml` with `val_period: 1` (validate every epoch by default)
- Updated the training loop in `ultralytics/engine/trainer.py` to run validation only when needed 🔁
  - Validation now runs when `val=True` **and**:
    - the current epoch matches the `val_period` schedule, **or**
    - it’s the final epoch, **or**
    - early stopping may trigger (`possible_stop`), **or**
    - training is stopping

### 🎯 Purpose & Impact
- Speeds up training workflows by reducing validation frequency on long runs 🏎️
- Helps users trade off between faster iteration and frequent metric updates 📉📈
- Keeps important safety checks: final-epoch and early-stopping-related validation still run even with larger `val_period` 🛑
- Fully backward-compatible: default behavior remains validation every epoch (`val_period=1`) 🔒

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Bovey0809, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message, and an Ultralytics engineer will assist you shortly. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -146,6 +146,7 @@ def __init__(self, cfg=DEFAULT_CFG, overrides=None, _callbacks=None):
             YAML.save(self.save_dir / "args.yaml", args_dict)  # save run args
         self.last, self.best = self.wdir / "last.pt", self.wdir / "best.pt"  # checkpoint paths
         self.save_period = self.args.save_period
+        self.val_period = self.args.val_period
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: ⚠️ **HIGH**: `self.args.val_period` may be missing (e.g., older `args.yaml` / overrides created before this key existed) or non-int (string from CLI/YAML). That would raise `AttributeError` here or `TypeError` later in the modulo check. Safer to `getattr` with a default and coerce to `int` once.

**Suggested change:**
```suggestion
        self.val_period = 1 if getattr(self.args, "val_period", None) is None else int(getattr(self.args, "val_period"))
```

File: ultralytics/engine/trainer.py [hunk 2]
```diff
@@ -473,7 +474,17 @@ def _do_train(self):
 
             # Validation
             final_epoch = epoch + 1 >= self.epochs
-            if self.args.val or final_epoch or self.stopper.possible_stop or self.stop:
+            # Run validation if: val=True AND (val_period matches OR final epoch OR early stopping)
+            should_val = (
+                self.args.val
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py]: 💡 **MEDIUM**: If `val_period` comes through as a non-int type (e.g., string), `(epoch + 1) % self.val_period` will fail. Casting `val_period` to `int` in `__init__` (see comment above) avoids runtime failures here, especially for CLI/YAML overrides.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Overall change is clean and the validation gating logic matches the stated intent (validate only when `val=True` and on the scheduled epochs / final / early-stop). One likely runtime issue: `val_period` may be absent or non-integer in some workflows (older saved configs or overrides), which can cause `AttributeError`/`TypeError`. Adding a defensive `getattr` + `int()` cast in `__init__` should make this robust.

💬 Posted 2 inline comments

**Comment by glenn-jocher** [comment]
@Bovey0809 this is a good idea! I tried to do this a while ago but ran into issues with saved results txt and png. We need to make sure that we report something every epoch for these, maybe -1 or NaN or something. Have you checked this out?

CC @Laughing-q

File: ultralytics/engine/trainer.py [hunk 3]
```diff
@@ -146,6 +146,7 @@ def __init__(self, cfg=DEFAULT_CFG, overrides=None, _callbacks=None):
             YAML.save(self.save_dir / "args.yaml", args_dict)  # save run args
         self.last, self.best = self.wdir / "last.pt", self.wdir / "best.pt"  # checkpoint paths
         self.save_period = self.args.save_period
```
[the remark this answers is not in the document]
Comment (glenn-jocher) [reply ultralytics/engine/trainer.py:150]: Yes good point, @Bovey0809 please do this, and also now that I think about it we need to update Docs too, especially macros training arguments

[the remark this answers is not in the document]
Comment (glenn-jocher) [reply ultralytics/engine/trainer.py:150]: https://github.com/ultralytics/ultralytics/blob/main/docs/macros/train-args.md

**Comment by Bovey0809** [comment]
@glenn-jocher Yes, I noticed issue in saving results.txt, the lines are not aligned with same columns, I'm trying to fix this.
